Fix opencl

katrinabrock opened the pull request:
Addresses several issues with the opencl tests:
- Model object does not correctly identify whether model was compiled with stan_opencl. Therefore, models were not recompiled within the tests when recompilation is needed. Resolved this by supplying a different exe_file for each tests so recompilation will always happen.
- Tests referenced a metadata attribute that was not defined. I've added checks for opencl related attributes that are set when opencl is functioning properly.
- Version test used a data input that was never created.

#### Submission Checklist

- [X] Run unit tests
- [X] Declare copyright holder and agree to license (see below)

#### Summary

Please describe the purpose of the pull request.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting
(this will be you or your assignee, such as a university or company):
Max Planck Institute of Animal Behavior


By submitting this pull request, the copyright holder is agreeing to
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

SteveBronder commented:
It sounds like the actual issue here is that cmdstanr does not correctly identify whether model was compiled with stan_opencl, should we be changing that inside of the model class instead of changing this test?

jgabry commented:
@SteveBronder did you see this draft PR which I think is related? https://github.com/stan-dev/cmdstanr/pull/1023 Perhaps these should be combined?

SteveBronder commented:
Yes I think it would be good to merge these two PRs together into one

katrinabrock commented:
Sounds good, I'll polish that one up and combine them.

Any idea if the opencl tests are actually running in CI (as in is `CMDSTANR_OPENCL_TESTS` set? If not, why not?

Also trying to get hold of a windows machine so I can repo the windows unit test failures. If there's any way to run these on ubuntu with docker or similar, would love to hear.

katrinabrock commented:
guessing the test failures are due to this: https://github.com/r-lib/actions/issues/217

katrinabrock commented:
Sorry for the multiple pings in a day....but I found another related deepish problem.

This test seems to be erroneously passing in the master branch: https://github.com/stan-dev/cmdstanr/blob/master/tests/testthat/test-threads.R#L162-L166 As in, **both** STAN_THREADS is not successfully getting set to `false` (due to this issue: https://github.com/stan-dev/cmdstan/issues/1293), **and** `mod$sample` is not successfully checking whether or not the binary has that flag set....so the test is passing even though it shouldn't be.

MRE (from master branch):
```
> mod <- cmdstan_model(
+   file.path(cmdstan_path(), "examples", "bernoulli", "bernoulli.stan"),
+   cpp_options = list(stan_threads = FALSE),
+   force_recompile = TRUE
+ )
Compiling Stan program...
> mod$cpp_options()$STAN_THREADS
[1] TRUE
```

The fix is that somewhere we need to convert falsy values at the R level to completely unset the ENV variable at the bash/cpp level. I'm not sure exactly where in the code it makes the most sense to do this, but I'll take a stab at it. Alternatively, we could inform users that they need to set `cpp_options = list(stan_threads = '')` or `cpp_options = list(stan_threads = NULL)`.

andrjohns commented:
> The fix is that somewhere we need to convert falsy values at the R level to completely unset the ENV variable at the bash/cpp level

No we shouldn't do anything here, since this is behaviour in Cmdstan - not cmdstanr. The options a user specifies should match those that are then set in `make/local`, otherwise a user would get different behaviour from the same options depending on the interface

katrinabrock commented:
ok, I'll update the test and the docs then. That's much easier. :-)

katrinabrock commented:
Sorry for the wall of text, but I need some input before moving forward.

----
I believe this PR working. It accomplishes:
- opencl test fixes described in original description
- fixes #765 (makes cmdrstan check compile args accurately)
- makes behavior, warnings, and docs more consistent with cmdstan's interpretation of flags

Does not solve
- No automatic recompile when provided cpp_options don't match binary
-----
Looking for input on two topics:

### 1. What should the behavior be when a model object is from an already compiled model that was compiled with different args than provided in the `cmdstan_model` function?

Some Options:
- Silently do nothing at creation, complain or fail at sampling (current behavior)
- Warn at creation, keep current behavior at sampling
- Automatically recompile at creation

More context: To me, auto-recompile makes the most sense from a user experience perspective. However, this is not the current behavior of `cmdstan` itself. Currently, `cmdstan` considers a model up to date even if build args have changed. Not sure if this is a bug or expected behavior that is not explicitly documented. [Most relevant doc I could find.](https://mc-stan.org/docs/cmdstan-guide/compiling_stan_programs.html#dependencies)

### 2. How should we standardize storage and checking of cpp_options/build args/compile info? Relatedly, what should be the output of `mod$cpp_options()`?

There are two aspects at issue:
- should the names be uppercase (as passed to cmdstan) or lowercase (consistent with R style)?
- when the flag is not set, should the value be `NULL`/empty/absent (as passed to cmdstan) or `FALSE`/`false` (as retrieved from cmdstan model `info`)?

Some Options:
- The status-quo is definitely bad. It is inconsistent on whether the names (keys) are uppercase or lowercase, leading to inaccurate check results.
- The version I have in this PR currently uses lowercase (per this comment: https://github.com/stan-dev/cmdstanr/issues/765#issuecomment-2274393175) and FALSE (consistent with the current output of `model_compile_info`. Note: this changes the user-facing output of user-facing `mod$cpp_options()`.
- Standardizing on uppercase would change the user-facing output of `mod$cpp_options()` less, (but still some in certain cases).
- We could set both uppercase and lowercase for maximum backward compatibility.
- An imo cleaner, but bigger change would be to consider `cpp_options` and `model_compile_info` as more separate, but related concepts. `cpp_options` being the R/pre-compile version, `model_compile_info` being the post-compile version. We could compare whether they are consistent with one another, but not explicitly assign one to the other as done currently here: https://github.com/stan-dev/cmdstanr/blob/8b72eacff40e630547484fef438ee2699f62d456/R/model.R#L275. This proposal would mean adding a mod$compile_info() method based on the binary associated with the model, and leave in place mod$cpp_options() based on the arguments the user provides. Then we could complain, error out, recompile, or what have you at various points if they're inconsistent with one another.

-----

katrinabrock commented:
One more wrinkle...prior to cmdstan [v2.27.0](https://github.com/stan-dev/cmdstan/releases/tag/v2.27.0), this "model info" does not exist. In that case, if the binary is already compiled, we have know way of knowing what flags were used.

For that matter...I think we are assuming throughout that the version of cmdstan the user currently has configured is the same as the version the binaries were compiled with. Currently, from what I can tell, we are not checking this assumption at all. We could check using model info when available, but before 2.27, all bets are off.

I'll have to think a bit about a clean way to resolve this. I'd love to hear if others have ideas.

jgabry commented:
@andrjohns I'm a bit swamped with work at the moment and haven't had a chance to look into this yet. Do you by chance have time to comment on this?

jgabry commented:
> @andrjohns I'm a bit swamped with work at the moment and haven't had a chance to look into this yet. Do you by chance have time to comment on this?

Actually I think @andrjohns is quite busy too. @rok-cesnovar or @SteveBronder do either of you have time to comment on this?

katrinabrock commented:
@jgabry @andrjohns I have a version I'm pretty happy with now. It stores what the user requested as cpp options and what the binary has configured separately, and (unless compile=FALSE) re compiles if there is a mismatch. I ran the CI on my fork and only the windows tests are failing. If someone on the team is able to give some feedback, especially on the change in functionality, that would be great! I'd love to get this in mergeable state. In any case, I will just use this version for my own purposes.

katrinabrock commented:
@mitzimorris @SteveBronder Here's my explanation of this change as promised:
# Initial Symptom
- In the status quo, when a model object is creating with an existing exe_file, the build args for this binary are essentially ignored (even though they are collected with the `info` command). In this case, when `sample` (or similar) is run, it does arg checking with inaccurate information about whether `STAN_OPENCL` and `STAN_THREADS` are set. This results in both false-positive and false-negative erroring and warning. (Error when everything is fine and no error when something is wrong.) Simple example here: https://github.com/katrinabrock/cmdstanr/commit/76973789c5f85031ffca4fce028d69db9609f9b8#diff-f9bf409eca909f7147a92f3deaf3fdeb15bc4590ef2e6ad46caf9bfedf863160R127 this testcase fails in the status quo.

# My Changes
### User-facing Changes in `initialize`and `compile` logic
- If an exe_file exists when `initialize` is called and a recompile does not occur, cmdstanr uses the `info` command to get as much info as possible about the version and cpp options used to create this binary. When `sample` and other methods are subsequently run, it uses this info to determine if the args provided to `sample` (or sister method) are compatible with cpp options that were provided at compile time. (This is the main thing that was not working before.)
- If an exe_file exists when `initialize` is called and the user provided `cpp_options` that do not match the output of `info`, a recompile is triggered. (Unless `compile=FALSE`, in that case, there is just a warning).
- If the user provides `cpp_options` to `initialize` or `compile`, they are "remembered" for subsequent compilation/recompilations.

### Other User-facing Changes
- For error and warning logic that involves checking `cmdstan` version. `cmdstanr` no longer assumes that the version it has currently configured is the same as the version the model was compiled with. Instead, it uses the output of `info` if available. If `info` is unavailable, it uses the version currently configured, but in that case only generates warnings, not errors.
   - This change is only applied to cpp_option related functions.
- `mod$cpp_options()` give a deprecation warning as it was a mixture between user provided cpp options and the output of `info`. I attempted to leave the behavior of this function unchanged. Did not thoroughly test the unchanged-ness.
- `mod$exe_info()` gives the data from the `info` command as an R object.
  - With `update=TRUE` the command is re-run before returning. The user should only need to explicitly add `update=TRUE` if they re-compiled outside of that model object (e.g. by running `cmdstan` directly).
- `mod$precompiled_options()` gives the options that will be used if `compile()` method is run without explicitly passing `cpp_options`. If the user has provided `cpp_options` in past runs of `initialize` or `compile`, it will be those. `names` are standardized to lowercase. Values are unchanged.
- If the user provides `FALSE` on an option we know is a flag, warn them that this will result in the flag being set. (e.g. `list(opencl=FALSE)` turns opencl ON).

### Developer-facing changes to CmdStanModel
- Previously `private$precompile_cpp_options_` only held the user-specified options until a R-triggered compile takes place, then it would be set to null. Now, we keep this information around whether or these options have been applied to the binary.
- Previously `private$exe_file_` was only set if the exe_file exists (either passed by the user or in an R-triggered compile). Now, if the exe_file does not exist, this variable will still be set to the path that we think it will be written to.
- `private$exe_info_` new attribute separate from `cpp_options_` and `precompile_cpp_options_`. Holds info gleaned from `cmdstan`'s post-compile `info` command.

### Test related changes

- added `with_mock_cli`, `expect_mock_compile`, and `expect_no_mock_compile` that allow developers to test the `cmdstanr` recompile logic that with developer-defined outputs of `cmdstan`. This both enables a cleaner separation between `cmdstan` logic and `cmdstanr` logic and also allows tests of recompile logic to run much faster.
- since logic around stan features now uses model-specific stan version where possible, `fake_cmdstan_version` can now update this model-specific version as well as the package-level version
- For a bunch of tests that are not intended to test recompile logic, I added `recompile=TRUE`. These tests are intended to test something that happens after compilatio. Sometimes (even with my changes) a binary is leftover from a previous test and a recompile was not properly triggered. This resulted in a mismatch between the binary and the model object.

SteveBronder commented:
Thank you! These changes totally make sense now and handle a lot of the issues we have from dealing with a make build system. Monday I will give this a harder read through and have some Qs

andrjohns commented:
Apologies for the extended delay @katrinabrock! I'm catching back up on Stan dev this week, and have better access to an OpenCL system now. I'll put aside some time to review on Wednesday

SteveBronder commented:
@katrinabrock hi my apologies for the delay. Looking at the tests it looks like the new ones with compile options are failing?

katrinabrock commented:
Hi @SteveBronder,

Thanks for taking a look. That's correct. As mentioned in @bob-carpenter's comment above, right now, I'm not looking for approval of the code. I'd like to make sure that the functionality I'm trying to implement is the functionality desired by the core dev team.

Does everything I've listed in the comment above from three weeks ago seem like the right set of changes?

 If so, I'll go ahead and diagnosis and resolve the windows issue. If you would like me to make a different set of logic changes, I'd like to settle on what the logic should be and implement that before worrying about the cross platform aspect.

jgabry commented:
@SteveBronder when you have a chance do you mind taking another look, based on @katrinabrock's most recent comment. I think you're much more familiar with all these CmdStan C++ options than I am. It sounded like @katrinabrock was looking for confirmation that the outlined approach in https://github.com/stan-dev/cmdstanr/pull/1022#issuecomment-2379006621 is reasonable. It seems reasonable to me but it would be good to get either you (Steve) or @andrjohns to confirm too.

katrinabrock commented:
I think @mitzimorris could also look over the logic and see if there are any concerns since we would likely want to have cmdstanpy and cmdstanr in sync.

mitzimorris commented:
while CmdStanPy recognizes `cpp_options`

jgabry commented:
Thanks @mitzimorris for checking the logic. And thank you @katrinabrock for working on this!! It's great to have new contributors.

katrinabrock commented:
@mitzimorris Thanks for the input!

@jgabry Between you and mitzi, do you think this is enough approval to go forward with this logic? Or do we still need @SteveBronder or @andrjohns to look at it?

jgabry commented:
I trust @mitzimorris' judgement on this, so I think it's ok to proceed. 

It would be great if @SteveBronder can help review the PR when it's ready since he knows a lot more about these particular aspects of CmdStan than I do. @andrjohns would also be great if he has the time (I know he's been busy lately). 

Thanks again @katrinabrock for working on this!

SteveBronder commented on tests/testthat/helper-mock-cli.R:
```diff
@@ -0,0 +1,22 @@
+real_wcr <- wsl_compatible_run
+
+with_mocked_cli <- function(code, compile_ret, info_ret){
```
@jgabry do we use `=` or `<-` for assignment in cmdstanr?

jgabry replied:
`<-` (although I've found some `=` at times in the past, but I try to change them when I find them)

SteveBronder commented on tests/testthat/test-model-compile-user_header.R:
```diff
@@ -0,0 +1,104 @@
+  # Check recompilation upon changing header
+  file.create(file_that_exists)
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_no_mock_compile({
+    mod$compile(quiet = TRUE, user_header = tmpfile)
```
Sorry I'm just not familiar with `testthat`'s mocking schema. Is `mod` here the `mod` from line 36?

```
        mod <- cmdstan_model(
          stan_file = testing_stan_file("bernoulli_external"),
          exe_file = file_that_exists,
          user_header = tmpfile
        )
```

katrinabrock replied:
um....good question. I don't think so because that shouldn't be in scope anymore. Which begs the question "where is this `mod` object coming from. I'll look into it.

SteveBronder replied:
lol thank you. Odd!!

katrinabrock replied:
Looking close, I think you're guess was correct it is getting defined above. `with_mock` puts objects into `topenv()` so parent frame.

SteveBronder commented on tests/testthat/test-model-compile.R:
```diff
@@ -391,15 +395,20 @@ test_that("check_syntax() works with include_paths", {
   stan_program_w_include <- testing_stan_file("bernoulli_include")
 
   mod_w_include <- cmdstan_model(stan_file = stan_program_w_include, compile=TRUE,
-                                 include_paths = test_path("resources", "stan"))
+                                 include_paths = test_path("resources", "stan"),
+                                 force_recompile = TRUE)
```
What is the side effect at issue here? I am just reading the code right now but tomorrow should have time to run the tests

katrinabrock replied:
😬 That's a mystery I didn't solve completely. Some test above was causing a side effect. (I think I have a version of the code somewhere that marked which test that was.) I realized actually there could be many such cases that are not caused by my change, but revealed by my change. In this situation (and some others), I "fixed" the problem by added `force_recompile = TRUE` to the impacted test instead of actually cleaning up the culprit test.

katrinabrock replied:
Ok, based on looking at my git history for when I was attempting to debug this. This is the test causing this particular side effect:
https://github.com/stan-dev/cmdstanr/blob/master/tests/testthat/test-model-compile.R#L85

So if you uncomment force_recompile, there will be a failure. If you then comment out the above linked test, the later test will pass without recompilation. (If I recall correctly.)

SteveBronder replied:
Oh my. Okay it sounds like something is carrying state through the tests. I had another project yesterday to work on. Friday I can give this a whirl

SteveBronder commented on tests/testthat/test-model-compile.R:
```diff
@@ -496,15 +505,22 @@ test_that("cpp_options work with settings in make/local", {
   mod <- cmdstan_model(file)
-  expect_true(mod$cpp_options()$STAN_THREADS)
+  expect_true(
+    expect_warning(mod$cpp_options()$stan_threads, "Use mod\\$exe_info()")
```
Does `expect_warning on its own not work here?

jgabry replied:
I guess using both is testing that there's a warning and also that the value is `TRUE`?

katrinabrock replied:
@jgabry exactly. It depends what you want to test. `expect_warning` checks if a warning is emitted. `expect_true` tests the return value. (Or at least that's the intention. I can double-check that `expect_warning` returns the value of the expression and not the result of the check.

jgabry replied:
I'm pretty sure you're right that it does, so this should be OK.

SteveBronder reviewed (changes requested):
Apologies for my slow review! I have a few design questions below. But overall this is a very nice fix for the issue. I'll continue with a bit of review tomorrow. I still need to read over a lot of the tests

jgabry commented on R/model.R:
```diff
@@ -248,7 +252,7 @@ CmdStanModel <- R6::R6Class(
         private$stan_code_ <- readLines(stan_file)
         private$model_name_ <- sub(" ", "_", strip_ext(basename(private$stan_file_)))
-        private$precompile_cpp_options_ <- args$cpp_options %||% list()
+        private$precompile_cpp_options_ <- validate_precompile_cpp_options(args$cpp_options) %||% list()
```
We define our own version of this in utils.R so should be fine already

jgabry commented on tests/testthat/test-model-compile-user_header.R:
```diff
@@ -0,0 +1,104 @@
+
+  # Check recompilation upon changing header
+  file.create(file_that_exists)
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_no_mock_compile({
```
We haven't been running any linters on the code, although we could start doing that. Either way, I agree about keeping the lines short. There are likely some other places in the code where that needs to be fixed, although I've tried to keep them under 80 in the code I've written.

katrinabrock replied:
yeah, since it doesn't seem like a linter was run in the past, I didn't run one on my code either, but that makes it surely inconsistent. I think there is a way to run linter just on the diff so at least we don't make the problem worse. I can look into it.

jgabry replied:
> I think there is a way to run linter just on the diff so at least we don't make the problem worse. I can look into it.

If that's simple to do then that sounds good, otherwise don't worry about it. We can (and probably should) start running a linter on the whole package, maybe after merging this PR.

katrinabrock replied:
TBH - I wouldn't advise that. It basically creates a huge diff so your subsequent blames become less meaningful. I think there is a way to exclude certain commits from blame...but....

jgabry replied:
Hmm yeah that’s a good point

SteveBronder replied:
We did this years ago in the stan math library. We essentially have a huge file that tells the compiler to ignore any auto changes.

For cmdstanr we could set up a hook so that the linter runs before any future commits. So we would put the one global change in the `.git-blame-ignore-revs` then have a git hook that runs precommit to run the linter

https://github.com/stan-dev/math/blob/develop/.git-blame-ignore-revs

But again, seperate issue from the PR

@katrinabrock if you can try to be mindful of not having too long of lines that would be great, but no need to go through counting column

katrinabrock replied:
I went ahead and linted all my changes.

jgabry commented:
> Apologies for my slow review! I have a few design questions below. But overall this is a very nice fix for the issue. I'll continue with a bit of review tomorrow. I still need to read over a lot of the tests

Thanks @SteveBronder, much appreciated! And thanks in advance for continuing tomorrow.

jgabry commented on R/model.R:
```diff
@@ -2375,7 +2538,51 @@ model_variables <- function(stan_file, include_paths = NULL, allow_undefined = F
+
+
+is_variables_method_supported <- function(mod) {
+  cmdstan_version() >= "2.27.0" && mod$has_stan_file() && file.exists(mod$stan_file())
```
It does understand it actually (perhaps surprisingly!), so this should be ok

katrinabrock replied:
It's simply string comparison so there are limitations like this:
```
> '3.11.0'> '3.2.0'
[1] FALSE
```

Sure, we can make version comparison more robust. Do we want that to be part of this PR?

jgabry replied:
Oh good point! 

We could start using e.g. base R's `numeric_version()` function. I don't think it needs to be part of this PR unless you're feeling eager to tackle it, in which case feel free! Either way, I'll open an issue for this.

jgabry replied:
https://github.com/stan-dev/cmdstanr/issues/1037

katrinabrock replied:
This PR is already huge so I'll save that for another one.

katrinabrock commented on R/model.R:
```diff
@@ -2375,7 +2538,51 @@ model_variables <- function(stan_file, include_paths = NULL, allow_undefined = F
+  info <- list()
+  info_raw <- strsplit(strsplit(ret_stdout, "\n")[[1]], "=")
+  for (key_val in info_raw) {
+    if (length(key_val) > 1) {
+      key_val <- trimws(key_val)
+      val <- key_val[2]
+      if (!is.na(as.logical(val))) {
+        val <- as.logical(val)
+      }
+      info[[tolower(key_val[1])]] <- val
+    }
```
I don't think we should be checking from a valid set of options because some of these flags are for dependencies of stan (e.g. Eigen) and it's non-trivial to get a complete list.

katrinabrock commented on R/model.R:
```diff
@@ -2375,7 +2538,51 @@ model_variables <- function(stan_file, include_paths = NULL, allow_undefined = F
+    }
+  }
+
+  info[["stan_version"]] <- paste0(info[["stan_version_major"]], ".", info[["stan_version_minor"]], ".", info[["stan_version_patch"]])
+  info[["stan_version_major"]] <- NULL
+  info[["stan_version_minor"]] <- NULL
+  info[["stan_version_patch"]] <- NULL
```
Pretty sure this logic is unchanged from existing. It's only `+` because of rearranging/renaming things. For legacy `$cpp_options()` I think we should leave as-is to avoid changing user-facing method, but for the new `$exe_info` we can design it any way we want since we're starting fresh.

SteveBronder replied:
Oh yes seeing that now ty

katrinabrock commented on tests/testthat/helper-mock-cli.R:
```diff
@@ -0,0 +1,22 @@
+}
+
+expect_mock_compile <- function(object, ...) expect_message(object, regexp = 'mock-compile-was-called', ...)
+expect_no_mock_compile <- function(object, ...) expect_no_message(object, message = 'mock-compile-was-called' , ...)
```
Overall what I'm trying accomplish here is basically mimic python's `assert_called` and `assert_not_called`. (https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called).

**Logic is**: `expect_mock_compile` will pass if mock_compile is called (at all, doesn't matter how many times) and fail if mock_compile is never called. `expect_no_mock_compile` is the inverse. It passes if mock_compile is *not* called at all and fails if mock_compile is called (even once).

**Implementation is**: `mock_compile` emits a message with the contents `mock-compile-was-called`. `expect_mock_compile` checks for this message: passes if it detects such a message, fails if it does not. `expect_no_mock_compile` fails if a message with exactly this text is detected and passes otherwise. A message with any other text does not impact `expect_no_mock_compile`.

There is probably a more elegant way to achieve this. Open to suggestions.

jgabry replied:
> There is probably a more elegant way to achieve this. Open to suggestions.

There may be, but I can't think of one at the moment. I would go ahead with this implementation, unless @SteveBronder has a better idea.

SteveBronder replied:
That's fine thanks for the clarification. Can you add a little comment (the above) just so people in the future can know what to expect?

katrinabrock replied:
✅

katrinabrock commented:
@SteveBronder Thanks for the review. I responded to the comments where I had something to add from memory. Most of them either I will update the code based on your suggestion or I need to look a little closer and rerun the code to properly answer you question. It will probably be a few weeks before I can devote a chunk of time to updating everything.

With the mocks, I think it will make sense if you play around with them. They are doing what I want, but certainly can be designed better.

jgabry commented:
> Most of them either I will update the code based on your suggestion or I need to look a little closer and rerun the code to properly answer you question.

Sounds good. Thanks again for working on this!

jgabry commented on R/model.R:
```diff
@@ -2243,40 +2335,111 @@ CmdStanModel$set("public", name = "expose_functions", value = expose_functions)
+  )
+  for (flag in flags_set_if_defined)   {
+    if (isFALSE(cpp_options[[flag]])) warning(
+      toupper(flag), " set to ", cpp_options[flag], " Since this is a non-empty value, ",
+      "it will result in the corresponding ccp option being turned ON. To turn this",
+      " option off, use cpp_options = list(", flag, " = NULL)."
```
I'm not entirely sure what to do here. I think I understand why @andrjohns is hesitant to override the user input, but I agree it's confusing! If we're going to keep passing `FALSE` here then I agree with @katrinabrock that a warning makes sense.

SteveBronder replied:
Thanks for linking to the comment.

@andrjohns my thought is that, to users of our make files, yes that would be different behavior. But to an R user I think they would expect setting false here to turn off the feature, no? Also I think not having edge case rules makes it a little easier to program over

SteveBronder commented:
All good and thank you for the PR! I should have time Friday to mess with the test issues

katrinabrock commented on R/model.R:
```diff
@@ -230,10 +231,13 @@ CmdStanModel <- R6::R6Class(
     include_paths_ = NULL,
     using_user_header_ = FALSE,
-    precompile_cpp_options_ = NULL,
+    precompile_cpp_options_ = list(),
```
Should be fine to have `NULL` here. My `validate_precompile_cpp_options` function converts it to an empty list anyway. If you like, I can change even that to preserve `NULL` and ensure downstream functions can handle NULL.

jgabry replied:
Unfortunately I don't think we've been entirely consistent about this for internal code. User-facing default arguments should be `NULL` if we're just defaulting to whatever CmdStan does, but internal code like this I think we've used both `NULL` and empty lists. In fact it looks like `stanc_options_` and and `cpp_options_` use `list()` but the `precompile_*` ones use `NULL` on the current master branch. So I think using `list()` is fine, but maybe then make all the `precompile_*` ones use `list()` if possible. Either way, not a big deal.

katrinabrock replied:
I went ahead and switched them to NULL.

katrinabrock commented on R/model.R:
```diff
@@ -270,22 +274,43 @@ CmdStanModel <- R6::R6Class(
+          }
+          self$exe_file(cmdstan_ext(strip_ext(exe_base)))
+          if (dir.exists(self$exe_file())) {
+            stop("There is a subfolder matching the model name in the same folder as the model! Please remove or rename the subfolder and try again.", call. = FALSE)
```
This code block is hit when either no stan file is provided or compile is set to false.

If there's a folder that has the name that the executable is suppose do to have...indeed we have no executable and also we can't compile to create one because the folder is in the way.

If no stan file is provided...basically that means the user just gave us an exe_file path, and that path is a path to a directory, not an executable, so there's no way we can recover from that. We don't have a binary or a stan file.

If compile is set to false but a stan file is provided...this is more like a warning that both operations that call the binary will fail because there's no binary there. Also, compiling will fail because we can't put the binary where it belongs.


I believe with the status quo, this issue is ignored when the object is instantiated, and we just with for the user to try compiling or sampling or what have you and then fail at that point.

jgabry replied:
Thanks, that sounds reasonable. 

> I believe with the status quo, this issue is ignored when the object is instantiated, and we just with for the user to try compiling or sampling or what have you and then fail at that point.

I think that's probably right

katrinabrock replied:
If you agree with my reasoning, I think no changes are needed to my PR. If you disagree, please un-resolve this comment.

katrinabrock commented on R/model.R:
```diff
@@ -328,9 +353,72 @@ CmdStanModel <- R6::R6Class(
     },
+    exe_info = function(update = FALSE) {
+      if (update) {
+        if (!file.exists(private$exe_file_)) return(NULL)
```
Yes and no.

The user doesn't necessarily have an awareness of what actions will cause exe_info(update=TRUE) to be called. Raising an error here would probably be more confusing than helpful to the user.

When the user provides a stan file, no binary exists, and sets compile=FALSE, in that case, I think the existing behavior is desirable. The user knows they haven't compiled yet and so it make sense that exe_info would be NULL.

User either doesn't pass a stan file or sets compile=FALSE, but no binary exists, in that case, I think a warning would make sense. I don't think there is one currently. However, I still wouldn't put that warning inside this method.

If the user deleted the binary at some point after instantiating the R object, I think we have no obligation to warn them. Philosophically, if the user is messing around with the filesystem behind cmdstanr's back. They do it at their own risk and we don't protect against all the strange and bad states that can result. (There are a lot of such states that are possible.)

Let me know if you'd like me to change anything.

jgabry replied:
Thanks @katrinabrock, that makes sense to me. @SteveBronder what do you think?

katrinabrock commented on R/model.R:
```diff
@@ -478,14 +567,20 @@ compile <- function(quiet = TRUE,
+
+  if (!is.null(user_header) && (
+    !is.null(cpp_options[["USER_HEADER"]]) || !is.null(cpp_options[["user_header"]])
+  )) warning("User header specified both via user_header argument and via cpp_options arguments")
```
Argument has highest precedence. If both user_header and USER_HEADER are specified in in the list, whichever is listed first is used regardless of capitalization.

katrinabrock commented on R/model.R:
```diff
@@ -2375,7 +2538,51 @@ model_variables <- function(stan_file, include_paths = NULL, allow_undefined = F
+      command = wsl_safe_path(exe_file),
+      args = "info",
+      error_on_status = FALSE
+    )
```
done

jgabry commented on R/model.R:
```diff
@@ -2243,40 +2335,111 @@ CmdStanModel$set("public", name = "expose_functions", value = expose_functions)
+  }
+
+  names(cpp_options) <- tolower(names(cpp_options))
+  flags_set_if_defined <- c(
+    # cmdstan
+    "stan_threads", "stan_mpi", "stan_opencl", "stan_no_range_checks", "stan_cpp_optims",
+    # stan math
+   "integrated_opencl", "tbb_lib", "tbb_inc", "tbb_interface_new"
+  )
```
Thanks, if it's just two hardcoded lists that don't fully overlap then I think your current approach is OK!

katrinabrock commented on R/model.R:
```diff
@@ -270,22 +274,43 @@ CmdStanModel <- R6::R6Class(
-              (!is.logical(cpp_options[[cpp_option_name]]) || isTRUE(cpp_options[[cpp_option_name]]))) {
-            private$cpp_options_[[cpp_option_name]] <- cpp_options[[cpp_option_name]]
+      } else {
+        # set exe path, same logic as in compile
```
✅ Done

katrinabrock commented:
@SteveBronder @jgabry I think I've addressed everything from @SteveBronder's last review. Based on his last comment, I think that review was partial and he wanted to look at the tests a bit more.

Based on the comments thus far, seems like the overall approach is sound so it will be worthwhile to fix the windows issue. I haven't done that yet, but I plan to.

andrjohns reviewed (commented):
Thanks for all of this work @katrinabrock! Given the amount of review that's already happened and how long it's taken me to look at it, I'm not going to hold this up with any substantive requests.

Just a minor style change, can you update all of the single-line if statements to be split across multiple lines? For example:

```r
 if (!file.exists(private$exe_file_)) return(NULL)
```
to:
```r
 if (!file.exists(private$exe_file_)) {
   return(NULL)
 }
```

I'll work through any windows/WSL issues now

bob-carpenter commented:
@andrjohns:  Is there an automatic formatter for R we can configure to do this automatically?  We got tired of having our PRs dominated by formatting, so we now use black for Python and ClangFormat for C++.

katrinabrock commented:
Looks like the unit tests are failing now. Even on linux 🤔 .

### Important part

@andrjohns I can probably solve the windows issue myself. I'd rather you spend your time looking at the functionality and implementation. I'm a bit concerned that I haven't really gotten substantive critiques yet. Steve's comments were mostly clarifications, formatting, and small issues. (Maybe because the substance is already perfect? 🤷‍♀️ 😁 😈 ) Those kinds of comments are valuable, I'd just like to make sure that some folks on the core dev team understand the changes deeply enough that they'll be patched instead of reverted if some small bug reports come in after rollout. I'm certainly willing to help, but I don't want to be the only one who understands this now 1.1K line piece of the codebase.  

btw - I'm still open to re-breaking this up into a bunch of smaller PRs.

### Less important part

@bob-carpenter Yes, `lintr` package can search for issues and `styler` package will fix them for you. 
@andrjohns Here is the `.lintr` file that I used to check my diff. (I fixed manually because I only wanted to touch the lines I was already changing.)

```
linters: linters_with_defaults(
    line_length_linter(85),
    object_usage_linter = NULL
  )
```

Currently, afaik there are no default linters that even allow you to set a requirement disallowing single line `if`. I think if this is something you'd like enforced in your project, you should build a custom linter for it. I can do a Q&D version of it for this PR.

jgabry commented:
>  I'd just like to make sure that some folks on the core dev team understand the changes deeply enough that they'll be patched instead of reverted if some small bug reports come in after rollout. I'm certainly willing to help, but I don't want to be the only one who understands this now 1.1K line piece of the codebase.

Especially since the C++ options stuff isn't my forte, I agree it would be great for @SteveBronder and @andrjohns to also understand the changes in this PR well enough (maybe they already do, not sure!) to do patches in the future instead of reverting if some bugs come up.

katrinabrock commented:
@SteveBronder @andrjohns Thinking it might be easier to review if I break this PR into multiple, smaller PRs...especially because I made multiple refactor-type changes.

First small PR is here:
https://github.com/stan-dev/cmdstanr/pull/1078

Hopefully this makes review more doable. 🙏

katrinabrock commented:
Second small PR: https://github.com/stan-dev/cmdstanr/pull/1079

These two PRs are independent and can be merged in either order.

katrinabrock commented:
Third small PR:
 - #1080

katrinabrock commented:
Fourth Small PR:
- https://github.com/stan-dev/cmdstanr/pull/1084

katrinabrock commented:
Fifth small PR:
- https://github.com/stan-dev/cmdstanr/pull/1085

jgabry commented:
> I plan to continue picking off pieces that can be merged on their own bit by bit.

That sounds great. 

> Mac - Release passing, devel failing - not sure if this failure is related to my changes

Not 100% sure, but I'm pretty confident this particular failure is one that I've seen happen occasionally for mysterious reasons (and then go away on subsequent runs of the GHA workflow) and is likely unrelated to your changes.

jgabry commented:
After rerunning the tests everything is passing now except on Windows
